definitions/types/gcp-alarm-metric.json

```diff
@@ -0,0 +1,75 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema",
+  "type": "object",
+  "title": "GCP Alarm Metric",
+  "required": [
+    "metric_type",
+    "resource_type",
+    "threshold",
+    "alignment_period_s",
+    "duration_s",
+    "value_type"
+  ],
+  "properties": {
+    "metric_type": {
+      "type": "string",
+      "title": "Metric Type",
+      "examples": [
+        "redis.googleapis.com/stats/memory/usage_ratio"
+      ]
+    },
+    "resource_type": {
+      "type": "string",
+      "title": "Resource Type",
+      "description": "The GCP Monitoring [Resource Type](https://cloud.google.com/monitoring/api/resources).",
+      "examples": [
+        "redis_instance"
+      ]
+    },
+    "threshold": {
+      "type": "number",
+      "title": "Threshold",
+      "description": "The threshold value for the alarm, between 0 and 1.",
+      "examples": [
+        0.9
+      ],
+      "minimum": 0,
+      "maximum": 1
+    },
+    "alignment_period_s": {
+      "type": "integer",
+      "title": "Period",
+      "description": "The alignment period for per-time series alignment",
+      "default": 60,
+      "minimum": 60,
+      "maximum": 600
+    },
+    "duration_s": {
+      "type": "integer",
+      "title": "Duration",
+      "description": "The amount of time that a time series must fail to report new data to be considered failing (in seconds). Only multiples of full minutes are supported.",
+      "default": 60,
+      "enum": [
+        60,
+        120,
+        180,
+        240,
+        300,
+        360,
+        420
+      ]
+    },
+    "value_type": {
```

Comment:
does this require a user to know something about the underlying metric?

Reply:
Also does string make this difficult to work with?

Reply:
It does yes. I made this an enum to make the choices valid, but like the `metric_type`, this too is metric-specific. I think we can get around some of the pain with defaults but it makes the massdriver yaml long again.

```
monitoring:
      allOf:
        - $ref: ../artifact-definitions/definitions/types/gcp-monitoring.json
        - dependencies:
            mode:
              oneOf:
                - properties:
                    mode:
                      const: CUSTOM
                    alarms:
                      type: object
                      title: Alarms
                      properties:
                        cpu_utilization:
                          $ref: ../artifact-definitions/definitions/types/gcp-alarm-metric.json
                          title: CPU Utilization
                          properties:
                            # for each metric, we can set constants and only allow things like the threshold to be adjusted
                            value_type:
                              const: "BOOL"
```

Reply:
It crossed my mind, why not make `alarms` top-level. We could generate the `_[alarms|monitoring].tf` from there... and then any `ServiceMonitor` yaml or ts files for other provisioners. Future alarms are yaml-only. Also, `mass alarm [list|add]` etc... 🤷 

```
alarms:
  properties:
    cpu_utilization:
      $ref: ../artifact-definitions/definitions/types/gcp-alarm-metric.json
      properties:
        display_name:
          type: string
          const: CPU Utilization
        metric:
          type: object
          propterties:
            metric_type:
              const: redis.googleapis.com/stats/cpu_utilization
            resource_type:
              const: redis_instance
            threshold:
              const: 0.9s
    memory_utilization:
      $ref: ../artifact-definitions/definitions/types/gcp-alarm-metric.json
      properties:
        display_name:
          type: string
          const: Memory Utilization
        metric:
          type: object
          propterties:
            metric_type:
              const: redis.googleapis.com/stats/memory/usage_ratio
            resource_type:
              const: redis_instance
            threshold:
              const: 0.9s
```

Reply:
Then of course you could squish that into generic types for each resource.

```
alarms:
  properties:
    cpu_utilization:
      $ref: ../artifact-definitions/definitions/types/gcp-alarm-redis-cpu.json
    memory_utilization:
      $ref:  ../artifact-definitions/definitions/types/gcp-alarm-redis-memory.json 

```

Reply:
@xpositivityx thoughts?